File: code/jvm/repository-jdbi/src/main/java/pt/isel/repositories/jdbi/channels/ChannelMemberRepositoryJdbi.java
package pt.isel.repositories.jdbi.channels;

import org.jdbi.v3.core.Handle;
import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.channels.ChannelMember;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.channels.ChannelMemberRepository;
import pt.isel.repositories.jdbi.utils.JdbiUtils;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import java.util.Map;

public class ChannelMemberRepositoryJdbi implements ChannelMemberRepository {
    private final Handle handle;
    private final String baseQuery = """
            SELECT
                cm.id as cm_id, cm.access_type,
                member.id AS member_id, member.username AS member_username,
                c.id AS channel_id, c.name AS channel_name, c.is_public AS channel_is_public,
                owner.id AS owner_id, owner.username AS owner_username
            FROM channel_members cm
            JOIN users member ON cm.user_id = member.id
            JOIN channels c ON cm.channel_id = c.id
            JOIN users owner ON c.owner_id = owner.id
            """;

    public ChannelMemberRepositoryJdbi(Handle handle) {
        this.handle = handle;
    }

    @Override
    public ChannelMember addUserToChannel(UserInfo userInfo, Channel channel, AccessType accessType) {
        String sql = """
                INSERT INTO channel_members (user_id, channel_id, access_type)
                VALUES (:user_id, :channel_id, :access_type)
                """;
        Long id = JdbiUtils.executeUpdateAndReturnId(handle, sql, JdbiUtils.params(
                "user_id", userInfo.id(),
                "channel_id", channel.id(),
                "access_type", accessType.name()
        ));
        return new ChannelMember(id, userInfo, channel, accessType);
    }

    @Override
    public ChannelMember findById(Long id) {
        return JdbiUtils.executeQueryToSingle(handle, baseQuery + " WHERE cm.id = :id",
                JdbiUtils.params("id", id), this::mapRowToChannelMember);
    }

    @Override
    public ChannelMember findUserInChannel(Long userId, Long channelId) {
        return JdbiUtils.executeQueryToSingle(handle,
                baseQuery + " WHERE cm.channel_id = :channel_id AND cm.user_id = :user_id",
                JdbiUtils.params("channel_id", channelId, "user_id", userId),
                this::mapRowToChannelMember);
    }

    @Override
    public List<ChannelMember> findAllChannelsForUser(Long userId, int limit, int offset) {
        return JdbiUtils.executeQueryToList(handle,
                baseQuery + " WHERE cm.user_id = :user_id ORDER BY cm.id ASC OFFSET :offset LIMIT :limit",
                JdbiUtils.params("user_id", userId, "offset", offset, "limit", limit),
                this::mapRowToChannelMember);
    }

    @Override
    public List<ChannelMember> findAllMembersInChannel(Long channelId, int limit, int offset) {
        return JdbiUtils.executeQueryToList(handle,
                baseQuery + " WHERE cm.channel_id = :channel_id ORDER BY cm.id ASC OFFSET :offset LIMIT :limit",
                JdbiUtils.params("channel_id", channelId, "offset", offset, "limit", limit),
                this::mapRowToChannelMember);
    }

    @Override
    public List<ChannelMember> findAll() {
        return JdbiUtils.executeQueryToList(handle, baseQuery, Map.of(), this::mapRowToChannelMember);
    }

    @Override
    public void save(ChannelMember entity) {
        JdbiUtils.executeUpdate(handle, """
                UPDATE channel_members
                SET access_type = :accessType
                WHERE id = :id
                """, JdbiUtils.params("id", entity.id(), "accessType", entity.accessType().name()));
    }

    @Override
    public void removeUserFromChannel(Long userId, Long channelId) {
        JdbiUtils.executeUpdate(handle, """
                DELETE FROM channel_members
                WHERE user_id = :user_id AND channel_id = :channel_id
                """, JdbiUtils.params("user_id", userId, "channel_id", channelId));
    }

    @Override
    public void deleteById(Long id) {
        JdbiUtils.executeUpdate(handle, "DELETE FROM channel_members WHERE id = :id", Map.of("id", id));
    }

    @Override
    public void clear() {
        JdbiUtils.executeUpdate(handle, "DELETE FROM channel_members", Map.of());
    }

    private ChannelMember mapRowToChannelMember(ResultSet rs) throws SQLException {
        UserInfo member = new UserInfo(rs.getLong("member_id"), rs.getString("member_username"));
        UserInfo owner = new UserInfo(rs.getLong("owner_id"), rs.getString("owner_username"));
        Channel channel = new Channel(rs.getLong("channel_id"), rs.getString("channel_name"), owner, rs.getBoolean("channel_is_public"));

        return new ChannelMember(
                rs.getLong("cm_id"),
                member,
                channel,
                AccessType.valueOf(rs.getString("access_type"))
        );
    }
}

File: code/jvm/repository-jdbi/src/main/kotlin/pt/isel/repositories/jdbi/ChannelMemberRepositoryJdbi.kt
package pt.isel.repositories.jdbi

import org.jdbi.v3.core.Handle
import pt.isel.domain.channel.AccessType
import pt.isel.domain.channel.Channel
import pt.isel.domain.channel.ChannelMember
import pt.isel.domain.user.UserInfo
import pt.isel.repositories.ChannelMemberRepository
import java.sql.ResultSet

class ChannelMemberRepositoryJdbi(
    private val handle: Handle,
) : ChannelMemberRepository {
    override fun addUserToChannel(
        userInfo: UserInfo,
        channel: Channel,
        accessType: AccessType,
    ): ChannelMember {
        val id =
            handle.executeUpdateAndReturnId(
                """
                INSERT INTO dbo.channel_members (user_id, channel_id, access_type)
                VALUES (:user_id, :channel_id, :access_type)
                """,
                mapOf(
                    "user_id" to userInfo.id,
                    "channel_id" to channel.id,
                    "access_type" to accessType,
                ),
            )
        return ChannelMember(id, userInfo, channel, accessType)
    }

    private val baseQuery = """
        SELECT
            cm.id as cm_id, cm.access_type,
            member.id AS member_id, member.username AS member_username,
            c.id AS channel_id, c.name AS channel_name, c.is_public AS channel_is_public,
            owner.id AS owner_id, owner.username AS owner_username
        FROM dbo.channel_members cm
        JOIN dbo.users member ON cm.user_id = member.id
        JOIN dbo.channels c ON cm.channel_id = c.id
        JOIN dbo.users owner ON c.owner_id = owner.id
    """

    override fun findById(id: Long): ChannelMember? =
        handle.executeQueryToSingle(
            "$baseQuery WHERE cm.id = :id",
            mapOf("id" to id),
            ::mapRowToChannelMember,
        )

    override fun findUserInChannel(
        userId: Long,
        channelId: Long,
    ): ChannelMember? =
        handle.executeQueryToSingle(
            "$baseQuery WHERE cm.channel_id = :channel_id AND cm.user_id = :user_id",
            mapOf("channel_id" to channelId, "user_id" to userId),
            ::mapRowToChannelMember,
        )

    override fun findAllChannelsForUser(
        userId: Long,
        limit: Int,
        offset: Int,
    ): List<ChannelMember> =
        handle.executeQueryToList(
            "$baseQuery WHERE cm.user_id = :user_id OFFSET :offset LIMIT :limit",
            mapOf("user_id" to userId, "offset" to offset, "limit" to limit),
            ::mapRowToChannelMember,
        )

    override fun findAllMembersInChannel(
        channelId: Long,
        limit: Int,
        offset: Int,
    ): List<ChannelMember> =
        handle.executeQueryToList(
            "$baseQuery WHERE cm.channel_id = :channel_id OFFSET :offset LIMIT :limit",
            mapOf("channel_id" to channelId, "offset" to offset, "limit" to limit),
            ::mapRowToChannelMember,
        )

    override fun findAll(): List<ChannelMember> =
        handle.executeQueryToList(
            baseQuery,
            mapper = ::mapRowToChannelMember,
        )

    override fun save(entity: ChannelMember) {
        handle.executeUpdate(
            """
            UPDATE dbo.channel_members
            SET access_type = :accessType
            WHERE id = :id
            """,
            mapOf("id" to entity.id, "accessType" to entity.accessType),
        )
    }

    override fun removeUserFromChannel(
        userId: Long,
        channelId: Long,
    ) {
        handle.executeUpdate(
            """
            DELETE FROM dbo.channel_members
            WHERE user_id = :user_id AND channel_id = :channel_id
            """,
            mapOf("user_id" to userId, "channel_id" to channelId),
        )
    }

    override fun deleteById(id: Long) {
        handle.executeUpdate("DELETE FROM dbo.channel_members WHERE id = :id", mapOf("id" to id))
    }

    override fun clear() {
        handle.executeUpdate("DELETE FROM dbo.channel_members")
    }

    private fun mapRowToChannelMember(rs: ResultSet): ChannelMember {
        val member =
            UserInfo(
                rs.getLong("member_id"),
                rs.getString("member_username"),
            )

        val owner =
            UserInfo(
                rs.getLong("owner_id"),
                rs.getString("owner_username"),
            )

        val channel =
            Channel(
                rs.getLong("channel_id"),
                rs.getString("channel_name"),
                owner,
                rs.getBoolean("channel_is_public"),
            )

        return ChannelMember(
            rs.getLong("cm_id"),
            member,
            channel,
            AccessType.valueOf(rs.getString("access_type")),
        )
    }
}
